validation_helper.go

```diff
@@ -0,0 +1,682 @@
+	for _, e := range i.definition.Enumerate {
+		if input == e {
+			return nil
+		}
+	}
+	return &IntEnumValidationError{
+		i.definition,
+		input,
+	}
+}
+
+type StringEnumValidator struct {
+	definition StringEnumValidatorDefinition
+}
+
+type StringEnumValidatorDefinition struct {
+	Enumerate []string `json:"enum"`
+}
+
+type StringEnumValidationError struct {
+	Definition StringEnumValidatorDefinition `json:"definition"`
+	Input      string                        `json:"input"`
+}
+
+func NewStringEnumValidator(definition StringEnumValidatorDefinition) (StringEnumValidator, error) {
+	enumerate := definition.Enumerate
+	if len(enumerate) == 0 {
+		return StringEnumValidator{}, EmptyError{"the enumerate should have at least one element"}
+	}
+
+	for i, e := range enumerate {
+		for _, es := range enumerate[i+1:] {
+			if e == es {
+				return StringEnumValidator{}, DuplicationError{"the elements of enumerate should not be duplicated"}
+			}
+		}
+	}
+	return StringEnumValidator{definition}, nil
+}
+
+func (s StringEnumValidator) Validate(input string) *StringEnumValidationError {
+	for _, e := range s.definition.Enumerate {
+		if input == e {
+			return nil
+		}
+	}
+	return &StringEnumValidationError{
+		s.definition,
+		input,
+	}
+}
+
+type FloatEnumValidator struct {
+	definition FloatEnumValidatorDefinition
+}
+
+type FloatEnumValidatorDefinition struct {
+	Enumerate []float64 `json:"enum"`
+}
+
+type FloatEnumValidationError struct {
+	Definition FloatEnumValidatorDefinition `json:"definition"`
+	Input      float64                      `json:"input"`
+}
+
+func NewFloatEnumValidator(definition FloatEnumValidatorDefinition) (FloatEnumValidator, error) {
+	enumerate := definition.Enumerate
+	if len(enumerate) == 0 {
+		return FloatEnumValidator{}, EmptyError{"the enumerate should have at least one element"}
+	}
+
+	for i, e := range enumerate {
+		for _, es := range enumerate[i+1:] {
+			if e == es {
+				return FloatEnumValidator{}, DuplicationError{"the elements of enumerate should not be duplicated"}
+			}
+		}
+	}
+	return FloatEnumValidator{definition}, nil
+
+}
+
+func (f FloatEnumValidator) Validate(input float64) *FloatEnumValidationError {
+	for _, e := range f.definition.Enumerate {
+		if input == e {
+			return nil
+		}
+	}
+	return &FloatEnumValidationError{
+		f.definition,
+		input,
+	}
+}
+
+type RequiredValidator struct {
+	definition RequiredValidatorDefinition
+}
+
+type RequiredValidatorDefinition struct {
+	Required []string `json:"pattern"`
+}
+
+type RequiredValidationError struct {
+	Input      interface{}                 `json:"input"`
+	Definition RequiredValidatorDefinition `json:"definition"`
+}
+
+func NewRequiredValidator(definition RequiredValidatorDefinition) (RequiredValidator, error) {
+	required := definition.Required
+	if len(required) == 0 {
+		return RequiredValidator{}, EmptyError{"the required value should have at least one element"}
+	}
+	for idx, e := range required {
```

Comment:
Wasted loop at the latest index, isn't it?

And the code needs more consistent. `idx` -> `i`